### PR TITLE
feat(proforge): responsive pipeline — abortable AI calls + determinate progress

### PR DESCRIPTION
**Pull request by qnbs**
## Wave B / PR7 — Responsive ProForge execution

The 8-stage ProForge pipeline is **network-bound** (each stage awaits AI calls, so the event loop is free during I/O) and already pauses between stages at the Human-in-the-Loop `awaitingReview` gate. Rather than a WorkerBus offload that gains little for network-bound work, this PR makes execution **responsive** along the three axes the plan calls for: cooperative yield, real progress, non-blocking abort.

### Changes
- **Non-blocking abort (real bug fix).** `generate()`/`selfReflect()` never populated `options.signal`, so aborting the pipeline waited for the in-flight AI request to finish. `BaseAgent.bindAbortSignal()` now threads the run's `AbortSignal` into every AI call (the provider already forwards `opts.signal` to `fetch`). Orchestrator **and** capability layer bind it before `execute()`, so abort is immediate and parity holds across entry points.
- **Cooperative yield.** `BaseAgent.cooperativeYield()` (a macrotask yield), adopted in the `copyEdit`/`prose` per-section loops, keeps the dashboard repainting during synchronous post-AI work (validation, dedup, repetition scans) on large manuscripts.
- **Determinate progress.** New pure `computePipelineProgress(run)` (`completed`/`total`/`percent`/`activeIndex`) + a `role="progressbar"` bar with "Stage N of M" / percent at the top of `PipelineProgressPanel`. i18n added across all 19 locales (core de/es/fr/it translated; Beta/RTL seeded for the bulk-MT pass).

Human-in-the-loop pause/review semantics and IDB run-history rehydration are untouched; abort stays cooperative.

### Tests
`pipelineProgress` math (8), signal-threading in `generate`/`selfReflect`, `cooperativeYield`, progress-bar render; orchestrator/capability mock agents gain `bindAbortSignal`. Local gate green: lint, typecheck, i18n parity, all touched proForge suites.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

___

## **CodeAnt-AI Description**
**Make ProForge progress clearer and cancellations take effect immediately**

### What Changed
- Added an overall progress bar in the ProForge panel with stage count and percent complete, so users can see how far the pipeline has advanced at a glance.
- When the active stage is not part of the selected stages, the panel now shows a neutral “Preparing…” label instead of a confusing stage number.
- Canceling a ProForge run now stops in-flight AI work right away, and aborted stages are reported as canceled instead of looking like a finished run.
- Copy edit, prose, and diagnostic runs now respect cancel requests across the pipeline, including during long section-by-section work.

### Impact
`✅ Clearer pipeline progress`
`✅ Faster cancel response`
`✅ Fewer mistaken “finished” aborts`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

**Comment by qnbs** [comment]
@codeant-ai review

File: services/proForge/proForgeCapabilityLayer.ts [hunk 1]
```diff
@@ -96,6 +96,8 @@ export class ProForgeCapabilityLayer {
       const agent = new AgentClass(context);
 
       const controller = new AbortController();
+      // QNBS-v3: PR7 — bind the signal so AI calls are cancellable (parity with the orchestrator).
+      agent.bindAbortSignal(controller.signal);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/proForge/proForgeCapabilityLayer.ts]: Fixed in fc777bdb. `runStage(rawInput, signal?)` now accepts a caller-provided AbortSignal and binds/passes that (falling back to a local controller only so the agent always has a signal). `runCopilotDiagnostic` forwards its signal, so Copilot/Node/MCP stage runs are genuinely cancellable. Test asserts the caller signal is bound to the agent.

File: components/proForge/PipelineProgressPanel.tsx [hunk 2]
```diff
@@ -53,6 +57,35 @@ export const PipelineProgressPanel: React.FC = () => {
...
     <div className="space-y-4">
+      {/* QNBS-v3: PR7 — determinate overall progress bar */}
+      <div className="rounded-sc-lg bg-[var(--sc-surface-elevated)] border border-[var(--sc-border-subtle)] p-4">
+        <div className="flex items-center justify-between mb-2">
+          <h3 className="text-sm font-medium">{t('proforge.progress.overall')}</h3>
+          <span className="text-xs text-[var(--sc-text-secondary)]">
+            {t('proforge.progress.stageOfTotal', {
+              current: progress.activeIndex,
+              total: progress.total,
+            })}
```
[the remark this answers is not in the document]
Comment (qnbs) [reply components/proForge/PipelineProgressPanel.tsx]: Fixed in fc777bdb. The panel now clamps the displayed stage to `Math.min(Math.max(activeIndex,1), total)` (0 only when no stages selected), so it never shows 'Stage 0 of M' when the active stage isn't in selectedStages. Added a panel test for the idle-active-stage edge case.

File: services/proForge/pipelineAgents/copyEditAgent.ts [hunk 3]
```diff
@@ -43,6 +43,9 @@ export class CopyEditAgent extends BaseAgent {
 
     for (const section of sectionsToProcess) {
       if (signal.aborted) break;
+      // QNBS-v3: PR7 — yield between sections so the UI stays responsive during synchronous
+      // post-AI processing (validation, dedup) on large manuscripts.
+      await this.cooperativeYield();
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/proForge/pipelineAgents/copyEditAgent.ts:48]: Fixed in fc777bdb. Added a second `if (signal.aborted) break;` immediately after `cooperativeYield()` so an abort during the yield can't let one more section + AI call through. Added a deterministic test (counting getter: top check passes, post-yield check aborts → generate never called).

File: services/proForge/pipelineAgents/proseAgent.ts [hunk 4]
```diff
@@ -64,6 +64,8 @@ export class ProseAgent extends BaseAgent {
 
     for (const section of sectionsToProcess) {
       if (signal.aborted) break;
+      // QNBS-v3: PR7 — yield between sections to keep the UI responsive during synchronous work.
+      await this.cooperativeYield();
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/proForge/pipelineAgents/proseAgent.ts:68]: Fixed in fc777bdb. Same post-yield `if (signal.aborted) break;` re-check added to the prose section loop, with a matching deterministic race test.

File: tests/unit/proForge/pipelineAgents/baseAgent.test.ts [hunk 5]
```diff
@@ -345,6 +348,31 @@ describe('BaseAgent', () => {
...
+      const opts = (mockGenerate.mock.calls.at(-1)?.[0] as { options: { signal?: AbortSignal } })
+        .options;
+      expect(opts.signal).toBe(controller.signal);
+    });
+  });
+
+  describe('cooperativeYield()', () => {
+    it('resolves on a later macrotask without throwing', async () => {
+      await expect(agent.publicCooperativeYield()).resolves.toBeUndefined();
+    });
```
[the remark this answers is not in the document]
Comment (qnbs) [reply tests/unit/proForge/pipelineAgents/baseAgent.test.ts]: Fixed in fc777bdb. The cooperativeYield test now uses fake timers and `vi.advanceTimersByTimeAsync(0)` to advance the macrotask explicitly — no reliance on real scheduling.

File: tests/unit/proForge/pipelineProgress.test.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    label: 'Test',
+    config: { ...DEFAULT_PIPELINE_CONFIG, selectedStages: ['intake', 'structural', 'lineProse'] },
+    status: 'running',
+    activeStage: 'structural',
+    stages: [],
+    startedAt: new Date(0).toISOString(),
+    prePipelineSnapshotId: 'snap-0',
+    traceLog: [],
+    ...overrides,
+  } as PipelineRun;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply tests/unit/proForge/pipelineProgress.test.ts]: Fixed in fc777bdb. `makeRun` is now annotated `: PipelineRun` and constructed without the `as` assertion, so the compiler verifies the full contract.

File: tests/unit/proForge/pipelineProgress.test.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    reviewItems: [],
+    metrics: {
+      aiCalls: 0,
+      tokensConsumed: 0,
+      durationMs: 0,
+      itemsFound: 0,
+      itemsAccepted: 0,
+      itemsRejected: 0,
+    },
+  } as unknown as PipelineRun['stages'][number];
```
[the remark this answers is not in the document]
Comment (qnbs) [reply tests/unit/proForge/pipelineProgress.test.ts]: Fixed in fc777bdb. The stage fixture is now `function stageResult(stage: PipelineStage, status: StageStatus): StageResult` returning a properly typed object — no `as unknown as`.

File: tests/unit/proForge/pipelineAgents/baseAgent.test.ts [hunk 8]
```diff
@@ -345,6 +348,31 @@ describe('BaseAgent', () => {
       expect(callArg.prompt).toContain('No structural edits found.');
       expect(callArg.prompt).toContain('Original prompt.');
     });
+
+    // QNBS-v3: PR7 — abort signal must reach the AI call so the request is cancellable mid-flight.
+    it('does NOT include a signal in options when none is bound', async () => {
+      mockGenerate.mockResolvedValueOnce({ text: 'OK', usage: {} });
+      await agent.publicGenerate('Prompt');
+      const opts = (mockGenerate.mock.calls.at(-1)?.[0] as { options: { signal?: AbortSignal } })
+        .options;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply tests/unit/proForge/pipelineAgents/baseAgent.test.ts]: Fixed in fc777bdb. Replaced the per-assertion cast with a typed `lastGenerateRequest(): GenerateRequest` accessor (the untyped mock's call payload assigns to GenerateRequest without an explicit assertion).

**Comment by qnbs** [comment]
@codeant-ai review

File: services/proForge/pipelineAgents/copyEditAgent.ts [hunk 9]
```diff
@@ -42,6 +42,12 @@ export class CopyEditAgent extends BaseAgent {
     let tokensConsumed = 0;
 
     for (const section of sectionsToProcess) {
+      if (signal.aborted) break;
+      // QNBS-v3: PR7 — yield between sections so the UI stays responsive during synchronous
+      // post-AI processing (validation, dedup) on large manuscripts.
+      await this.cooperativeYield();
+      // QNBS-v3: PR7 — re-check after the async yield: an abort during the yield must not let one
+      // more section (and its AI call) slip through.
       if (signal.aborted) break;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/proForge/pipelineAgents/copyEditAgent.ts:51]: Fixed in 7e7af7f0. `break` only exited the loop; copyEdit (and prose, same issue) now early-return on `signal.aborted` BEFORE the expensive post-loop work (repetition/filter-word scan, dedup, memory write, result assembly), so a cancelled run stops promptly and emits no stage output. Existing post-yield race tests already exercise this early return.

File: components/proForge/PipelineProgressPanel.tsx [hunk 10]
```diff
@@ -30,6 +31,14 @@ export const PipelineProgressPanel: React.FC = () => {
   const stageName = (stage: string): string => t(`proforge.stageName.${stage}`);
   const statusLabel = (status: string): string => t(`proforge.status.${status}`);
 
+  // QNBS-v3: PR7 — determinate overall progress (stage N of M / percent) for the active run.
+  const progress = useMemo(() => computePipelineProgress(currentRun), [currentRun]);
+  // QNBS-v3: PR7 — clamp the displayed stage number to a valid 1..total range; computePipelineProgress
+  // returns activeIndex 0 when the active stage isn't among the selected stages (e.g. idle/archived),
+  // which must never surface as "Stage 0 of M".
+  const displayStage =
+    progress.total > 0 ? Math.min(Math.max(progress.activeIndex, 1), progress.total) : 0;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply components/proForge/PipelineProgressPanel.tsx]: Fixed in 7e7af7f0. Reverted the clamp-to-1 — it misreported position. The panel now shows the 'Stage N of M' counter only when activeIndex >= 1 (active stage is actually selected); otherwise a neutral 'proforge.progress.preparing' label (new key, 19 locales). computePipelineProgress keeps returning activeIndex 0 for the not-found case.

File: tests/unit/proForge/pipelineAgents/baseAgent.test.ts [hunk 11]
```diff
@@ -40,10 +40,20 @@ vi.mock('../../../../services/logger', () => ({
...
 import type { OrchestratorContext } from '../../../../services/proForge/proForgeOrchestrator';
 
+// QNBS-v3: typed accessor for the last gateway.generate() call — avoids per-assertion `as` casts.
+// mockGenerate is an untyped vi.fn(), so calls[…][0] is `any` and assigns to GenerateRequest cleanly.
+function lastGenerateRequest(): GenerateRequest {
+  const calls = mockGenerate.mock.calls;
+  const last = calls[calls.length - 1];
+  if (!last) throw new Error('gateway.generate was not called');
+  return last[0];
+}
```
[the remark this answers is not in the document]
Comment (qnbs) [reply tests/unit/proForge/pipelineAgents/baseAgent.test.ts:57]: Fixed in 7e7af7f0. `lastGenerateRequest` now annotates the calls array as `GenerateRequest[][]` so the tuple element is typed GenerateRequest (no `any`-typed element, no `as`). Kept mockGenerate an untyped vi.fn() so it stays assignable to the InferenceGateway port (typing its return as unknown/GenerateResult broke that assignment or the existing minimal mocks).

**Comment by qnbs** [comment]
@codeant-ai review

File: services/proForge/pipelineAgents/copyEditAgent.ts [hunk 12]
```diff
@@ -97,6 +103,24 @@ export class CopyEditAgent extends BaseAgent {
...
+        metrics: {
+          aiCalls,
+          tokensConsumed,
+          durationMs: this.elapsed(startTime),
+          itemsFound: 0,
+          itemsAccepted: 0,
+          itemsRejected: 0,
+        },
+        agentOutput: undefined,
+      };
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/proForge/pipelineAgents/copyEditAgent.ts]: Fixed in 25894468. copyEdit/prose now THROW 'Stage aborted' before the post-loop work instead of returning an empty result — this both stops the wasted work (wave-2) and makes cancellation explicit so the capability-layer path can't read an aborted run as a successful empty completion. Additionally, runStage re-checks the signal after execute() and throws STAGE_FAILED, covering agents that complete without honouring the signal. Tests updated to expect rejection + a new capability aborted-signal test.

**Comment by qnbs** [comment]
@codeant-ai review
